### PR TITLE
build.d now builds on Windows x64

### PR DESCRIPTION
**Pull request by JinShil**
With these changes build.d can now build on Windows 64-bit.  At least on my computer.

I also added some more logging so I could see more of what was happening.  I don't understand the current "(DC) D_LEXER_OBJ" nomenclature, but I tried to follow it anyway.

There's more cleanup to be done to remove dependencies on Visual Studio and the environment, but I think I should do that in a followup PR.  I still need to find out if we can build DMD all in one shot, instead of building the backend and the frontend separately.

### CODE REVIEW

File: src/build.d [hunk 1]
```diff
@@ -721,6 +712,19 @@ void processEnvironment()
...
+        auto filename = path.baseName;
+
+        auto result = filename != "dt.d" && filename != "obj.d";
+        version(Posix)
+            result = result && filename != "machobj.d" && filename != "mscoffobj.d";
+        else version(Windows)
+            result = result && filename != "elfobj.d" && filename != "machobj.d";
+
+        return result;
+    }
```
Comment (wilzbach) [inline src/build.d]: The exclusion was just a temporary workaround. We can do this with version declarations at the top of these platform specific files - we have done this before for some of them and it allows us to get rid of this logic.

**Comment by wilzbach** [comment]
>  I still need to find out if we can build DMD all in one shot, instead of building the backend and the frontend separately.

While the technically possible, I would strongly advise against it as you typically don't need to rebuild the backend during development sessions and thus have considerably faster build times. It's also important that the backend builds without the frontend, because Walter uses it for DMC++ and on the other side, we need to ensure that LDC + GDC can build the frontend separately without any dependency on the Digitalmars backend.

File: src/build.d (hunk 1 above)
Comment (wilzbach) [reply src/build.d]: Example: https://github.com/dlang/dmd/blob/master/src/dmd/libmach.d

Comment (JinShil) [reply src/build.d]: I see now.  It appears that has already been done.  I've updated build.d and it seems to work.  Thanks.

File: src/build.d [hunk 2]
```diff
@@ -776,8 +767,8 @@ auto sourceFiles()
         backend:
             dirEntries(env["C"], "*.d", SpanMode.shallow)
                 .map!(e => e.name)
-                .filter!(e => !e.canFind("dt.d", "obj.d"))
-                .array ~ buildPath(env["C"], "elfobj.d"),
+                .filter!(e => e.baseName != "dt.d" && e.baseName != "obj.d")
```
Comment (jacob-carlborg) [inline src/build.d]: `optabgen.d` should be filtered out as well.

File: src/build.d [hunk 3]
```diff
@@ -776,8 +767,8 @@ auto sourceFiles()
         backend:
             dirEntries(env["C"], "*.d", SpanMode.shallow)
                 .map!(e => e.name)
-                .filter!(e => !e.canFind("dt.d", "obj.d"))
-                .array ~ buildPath(env["C"], "elfobj.d"),
+                .filter!(e => e.baseName != "dt.d" && e.baseName != "obj.d" && e.baseName != "optabgen.d")
```
Comment (wilzbach) [inline src/build.d]:
```suggestion
                .filter!(e => e.baseName.among("dt.d", "obj.d", "optabgen.d"))
```

Comment (JinShil) [reply src/build.d]: Should that be `!e.baseName.among`?

Comment (JinShil) [reply src/build.d]: I'll test locally and updat the PR.  Thanks.

File: src/build.d [hunk 4]
```diff
@@ -903,7 +886,7 @@ auto getHostCXXKind()
         return !cxxVersion.find("gcc", "Free Software")[0].empty ? "g++" : "clang++";
     }
     else version(Windows)
-        return "dmc";
+        return "";
```
Comment (wilzbach) [inline src/build.d]: Let's better purge this completely -> https://github.com/dlang/dmd/pull/9610

File: src/build.d [hunk 5]
```diff
@@ -247,7 +233,7 @@ version(Windows)
         Dependency dependency = {
             target: env["G"].buildPath(targetName).exeName,
             sources: [`vcbuild\` ~ targetName],
-            name: "(DC) MSCV-LIB " ~ targetName,
+            name: "(DC) MSCV-LIB",
```
Comment (wilzbach) [inline src/build.d]: Maybe:

```suggestion
            name: "(DC) MSCV-LIB " " ~ targetName.baseName,
```

Comment (JinShil) [reply src/build.d]: `targetName` is already the `baseName`, hence the need for `buildPath` for `target`.  Am I missing something?

Comment (JinShil) [reply src/build.d]: Maybe `targetName.exeName`?

Comment (wilzbach) [reply src/build.d]: Yeah whatever you prefer. All is more information than just MSCV-LIB

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**